cmd/boulder-wfe2/main.go

```diff
@@ -214,13 +237,56 @@ func loadCertificateChains(chainConfig map[string][]string, requireAtLeastOneCha
 
 		// Save the full PEM chain contents, if any
 		if buffer.Len() > 0 {
-			results[aiaIssuerURL] = buffer.Bytes()
+			results[id] = buffer.Bytes()
 		}
 	}
 
 	return results, issuerCerts, nil
 }
 
+// loadChain takes a list of filenames containing pem-formatted certificates,
+// and returns a chain representing all of those certificates in order. It
+// ensures that the resulting chain is valid. The final file is expected to be
+// a root certificate, which the chain will be verified against, but which will
+// not be included in the resulting chain.
+func loadChain(certFiles []string) (*issuance.Certificate, []byte, error) {
+	if len(certFiles) < 2 {
+		return nil, nil, errors.New(
+			"each chain must have at least two certificates: an intermediate and a root")
+	}
+
+	// Pre-load all the certificates to make validation easier.
+	certs := make([]*x509.Certificate, len(certFiles))
+	var err error
+	for i := 0; i < len(certFiles); i++ {
+		certs[i], err = core.LoadCert(certFiles[i])
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to load certificate: %w", err)
+		}
+	}
+
+	// Iterate over all certs except for the last, checking that their signature
+	// comes from the next cert in the list, and appending their pem to the buf.
+	var buf bytes.Buffer
+	for i := 0; i < len(certs)-1; i++ {
+		err = certs[i].CheckSignatureFrom(certs[i+1])
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to verify chain: %w", err)
+		}
+
+		buf.Write([]byte("\n"))
+		buf.Write(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certs[i].Raw}))
+	}
+
+	err = certs[len(certs)-1].CheckSignatureFrom(certs[len(certs)-1])
+	if err != nil {
+		return nil, nil, fmt.Errorf(
+			"final cert in chain must be a self-signed (used only for validation): %w", err)
```

Comment:
nit: the style in most Boulder packages is to capitalize the first letter of an error string. The [Go Code Review Guide](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) seems to dictate the opposite, so I am split on this.

Reply:
My understanding is we capitalize the first letter when
a) Sending something to the user; or
b) calling cmd.FailOnErr, because then the capital will be the first character of the whole error log message.

Here, where these errors will be wrapped again before they're surfaced anywhere, I prefer to stick with the go-recommended uncapitalized style. And although there certainly are many exceptions, we do use the uncapitalized style _mostly_:
```
❯ ag 'fmt.Errorf\("[A-Z]' | wc -l
134

…/boulder on wfe2-multi-issuer 
❯ ag 'fmt.Errorf\("[a-z]' | wc -l
848
```

Reply:
> My understanding is we capitalize the first letter when
> a) Sending something to the user; or
> b) calling cmd.FailOnErr, because then the capital will be the first character of the whole error log message.

This is a good rule of thumb, though I think the real answer is "it's mostly been chaos." Let's write down this rule in CONTRIBUTING.md and start following it.